pkg/storegateway/series_refs.go

```diff
@@ -1025,18 +1116,78 @@ func (s *loadingSeriesChunkRefsSetIterator) Err() error {
 }
 
 // loadSeries returns a for chunks. It is not safe to use the returned []chunks.Meta after calling loadSeries again
-func (s *loadingSeriesChunkRefsSetIterator) loadSeries(ref storage.SeriesRef, loadedSeries *bucketIndexLoadedSeries, stats *queryStats, builder *labels.ScratchBuilder) (labels.Labels, []chunks.Meta, error) {
-	ok, err := loadedSeries.unsafeLoadSeries(ref, &s.symbolizedLsetBuffer, &s.chunkMetasBuffer, s.skipChunks, stats)
+func (s *loadingSeriesChunkRefsSetIterator) loadSeries(ref storage.SeriesRef, loadedSeries *bucketIndexLoadedSeries, stats *queryStats, lsetPool *pool.SlabPool[symbolizedLabel]) ([]symbolizedLabel, []chunks.Meta, error) {
+	ok, lbls, err := loadedSeries.unsafeLoadSeries(ref, &s.chunkMetasBuffer, s.skipChunks, stats, lsetPool)
 	if !ok || err != nil {
-		return labels.EmptyLabels(), nil, errors.Wrap(err, "loadSeries")
+		return nil, nil, errors.Wrap(err, "loadSeries")
+	}
+
+	return lbls, s.chunkMetasBuffer, nil
+}
+
+func (s *loadingSeriesChunkRefsSetIterator) singlePassStringify(symbolizedSet symbolizedSeriesChunkRefsSet) (seriesChunkRefsSet, error) {
+	// Some conservative map pre-allocation; the goal is to get an order of magnitude size of the map, so we minimize map growth.
+	symbols := make(map[uint32]string, len(symbolizedSet.series)/2)
+
+	for _, series := range symbolizedSet.series {
+		for _, symRef := range series.lset {
+			symbols[symRef.value] = ""
+			symbols[symRef.name] = ""
+		}
+	}
+
+	allSymbols := make([]uint32, 0, len(symbols))
+	for sym := range symbols {
+		allSymbols = append(allSymbols, sym)
 	}
+	slices.Sort(allSymbols)
 
-	lset, err := s.indexr.LookupLabelsSymbols(s.symbolizedLsetBuffer, builder)
+	symReader, err := s.indexr.indexHeaderReader.SymbolsReader()
 	if err != nil {
-		return labels.EmptyLabels(), nil, errors.Wrap(err, "lookup labels symbols")
+		return seriesChunkRefsSet{}, err
+	}
+	defer symReader.Close()
+	for _, sym := range allSymbols {
+		symbols[sym], err = symReader.Read(sym)
+		if err != nil {
+			return seriesChunkRefsSet{}, err
+		}
+	}
+
+	set := newSeriesChunkRefsSet(len(symbolizedSet.series), symbolizedSet.releasable)
+
+	labelsBuilder := labels.NewScratchBuilder(16)
```

Comment:
Could we more accurately pre-allocate the size of this by computing the longest set of labels in the loop over `symbolizedSet.series` above (line 1132)? (and is it worth the effort?)

Reply:
i'm not super sure this is worth the effort. This scratch builder creates a new right-sized slice for each labels.Labels. Saving one slice resizing a few times probably doesn't make a difference. I can try to run some benchmakrs

Reply:
I don't know. I think running a benchmark will take longer. The change is small enough c2a5bc07c8472b59642f0a40235cc92290382592